src/Package/Unpack.zig

```diff
@@ -0,0 +1,382 @@
+        code: anyerror,
+        target_path: []const u8,
+        sym_link_path: []const u8,
+    },
+    unable_to_create_file: struct {
+        code: anyerror,
+        file_name: []const u8,
+    },
+};
+
+pub fn init(allocator: std.mem.Allocator, root: fs.Dir) Self {
+    return .{
+        .allocator = allocator,
+        .root = root,
+    };
+}
+
+pub fn deinit(self: *Self) void {
+    for (self.errors.items) |item| {
+        switch (item) {
+            .unable_to_create_sym_link => |info| {
+                self.allocator.free(info.target_path);
+                self.allocator.free(info.sym_link_path);
+            },
+            .unable_to_create_file => |info| {
+                self.allocator.free(info.file_name);
+            },
+        }
+    }
+    self.errors.deinit(self.allocator);
+    self.* = undefined;
+}
+
+const Self = @This();
+
+pub fn tarball(self: *Self, reader: anytype) !void {
+    const strip_components = 1;
+
+    var file_name_buffer: [std.fs.MAX_PATH_BYTES]u8 = undefined;
+    var link_name_buffer: [std.fs.MAX_PATH_BYTES]u8 = undefined;
+    var iter = std.tar.iterator(reader, .{
+        .file_name_buffer = &file_name_buffer,
+        .link_name_buffer = &link_name_buffer,
+    });
+    while (true) {
+        if (iter.next() catch |err| switch (err) {
+            error.TarUnsupportedHeader => continue,
+            else => return err,
+        }) |entry| {
+            switch (entry.kind) {
+                .directory => {}, // skip empty
+                .file => {
+                    if (entry.size == 0 and entry.name.len == 0) continue;
+                    const file_name = stripComponents(entry.name, strip_components);
+                    if (file_name.len == 0) return error.BadFileName;
+                    if (try self.createFile(file_name)) |file| {
+                        defer file.close();
+                        try entry.writeAll(file);
+                    }
+                },
+                .sym_link => {
+                    const file_name = stripComponents(entry.name, strip_components);
+                    if (file_name.len == 0) return error.BadFileName;
+                    const link_name = entry.link_name;
+                    try self.symLink(link_name, file_name);
+                },
+            }
+        } else break;
+    }
+}
+
+pub fn gitPack(self: *Self, commit_oid: git.Oid, reader: anytype) !void {
+    // Same interface as std.fs.Dir.createFile, symLink
+    const inf = struct {
+        parent: *Self,
+
+        pub fn makePath(_: @This(), _: []const u8) !void {}
+
+        pub fn createFile(t: @This(), sub_path: []const u8, _: fs.File.CreateFlags) !fs.File {
+            return (try t.parent.createFile(sub_path)) orelse error.Skip;
+        }
+
+        pub fn symLink(t: @This(), target_path: []const u8, sym_link_path: []const u8, _: fs.Dir.SymLinkFlags) !void {
+            try t.parent.symLink(target_path, sym_link_path);
+        }
+    }{ .parent = self };
```

Comment:
Agree.

That was also my thinking but since checkout is used in test and the main function I decided for the first step to make just minimal changes so those two can be mostly unchanged. 
Iterator which returns all entries, and does not require a recursive call for each directory, will be nice to use from the outside. 
That `error.Skip` is a really ugly hack.